src/bsp/platform/iotc_bsp_io_net_test.cc

```diff
@@ -0,0 +1,128 @@
+/* Copyright 2018-2019 Google LLC
+ *
+ * This is part of the Google Cloud IoT Device SDK for Embedded C,
+ * it is licensed under the BSD 3-Clause license; you may not use this file
+ * except in compliance with the License.
+ *
+ * You may obtain a copy of the License at:
+ *  https://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "iotc_test_echoserver.h"
+#include "gtest.h"
+#include "iotc_bsp_io_net.h"
+
+#include <stdio.h>
+
+namespace iotctest {
+namespace {
+
+typedef struct NetworkTestCase {
+  iotc_bsp_socket_type_t socket_type;
+  iotc_bsp_protocol_type_t protocol_type;
+  char* listening_addr;
+} NetworkType;
+
+class ServerTest : public ::testing::TestWithParam<NetworkType*> {
+  public:
+    const uint16_t kTestPort = 2000;
+    const char* kTestMsg = "hello\n";
+    bool WaitUntilSocketReadyForWrite();
+    bool WaitUntilSocketReadyForRead();
+    int out_written_count_, state_;
+    const char* listening_addr_;
+    iotc_bsp_socket_type_t socket_type_;
+    iotc_bsp_protocol_type_t protocol_type_;
+    iotc_bsp_socket_events_t socket_evts_[1];
+    iotc_bsp_socket_t test_socket_;
+};
+
+bool ServerTest::WaitUntilSocketReadyForWrite() {
+  bool ready_to_write = false;
+  memset(socket_evts_, 0, sizeof(iotc_bsp_socket_events_t));
+  socket_evts_[0].iotc_socket = test_socket_;
+  socket_evts_[0].in_socket_want_write = 1;
+  while (true) {
+    state_ = iotc_bsp_io_net_select(socket_evts_, 1, kTimeoutSeconds);
+    if (state_ == IOTC_BSP_IO_NET_STATE_OK &&
+        socket_evts_[0].out_socket_can_write == 1) {
+      ready_to_write = true;
+      break;
+    }
+  }
+  return ready_to_write;
+}
+
+bool ServerTest::WaitUntilSocketReadyForRead() {
+  bool ready_to_read = false;
+  memset(socket_evts_, 0, sizeof(iotc_bsp_socket_events_t));
+  socket_evts_[0].iotc_socket = test_socket_;
+  socket_evts_[0].in_socket_want_read = 1;
+  while (true) {
+    state_ = iotc_bsp_io_net_select(socket_evts_, 1, kTimeoutSeconds);
+    if (state_ == IOTC_BSP_IO_NET_STATE_OK &&
+        socket_evts_[0].out_socket_can_read == 1) {
+      ready_to_read = true;
+      break;
+    }
+  }
+  return ready_to_read;
+}
+
+TEST_P(ServerTest, EndToEndCommunicationWorks) {
+  const NetworkType* test_case = GetParam();
+  socket_type_ = test_case->socket_type;
```

Comment:
Make `socket_type_`, `listening_addr_`, `protocol_type_` local variables. There is no need them to be in the fixture class.